Add CSS color parsing example

minamorl opened the pull request.

wavebeem commented on examples/color.ts:
```diff
@@ -0,0 +1,133 @@
+    public g: RGBValue,
+    public b: RGBValue,
+    public a: Transparency | null
+  ) {}
```
I would like to focus more on the parser in this example, so maybe less class code?

Just the class Color like:

```ts
class Color {
  constructor(public r: number, public g: number, public b: number, public a: number = 1)  {}
  toString() { ... }
}
```

wavebeem commented on examples/color.ts:
```diff
@@ -0,0 +1,133 @@
+function repeat<T>(parser: bnb.Parser<T>, times: number) {
+  const parsers: bnb.Parser<T>[] = Array(times).fill(parser);
+  return bnb.all(...parsers);
+}
```
This is a cool parser, but I think this file would be easier to understand without it

wavebeem commented on examples/color.ts:
```diff
@@ -0,0 +1,133 @@
+// parsers
+const pattern1 = sharp
+  .next(repeat(hexval, 3))
+  .skip(bnb.eof)
```
I don't think you need this `.skip`

wavebeem commented on examples/color.ts:
```diff
@@ -0,0 +1,133 @@
+  .map((values) =>
+    values.map((v) => parseInt(v + v, 16)).map((v) => new RGBValue(v))
+  )
+  .map(([r, g, b]) => new Color(r, g, b, null));
```
These two `.map` can be combined like:

```js
.map(([r, g, b]) => {
  return new Color(
    parseInt(r + r, 16),
    parseInt(g + g, 16),
    parseInt(b + b, 16),
    1
  );
});
```

wavebeem commented on examples/color.ts:
```diff
@@ -0,0 +1,133 @@
+      values.map((v) => parseInt(v, 16)).map((v) => new RGBValue(v))
+    )
+    .map(([r, g, b]) => new Color(r, g, b, null))
+);
```
Same idea as above refactor

wavebeem commented on examples/color.ts:
```diff
@@ -0,0 +1,133 @@
+    ...[r, g, b].map((v) => new RGBValue(v)),
+    new Transparency(a),
+  ])
+  .map(([r, g, b, a]) => new Color(r, g, b, a));
```
Easier to read with a single `.map` call I think.

wavebeem commented on examples/color.ts:
```diff
@@ -0,0 +1,133 @@
+  ])
+  .map(([r, g, b, a]) => new Color(r, g, b, a));
+
+const parsers = bnb.choice(pattern1, pattern2, pattern3, pattern4);
```
Maybe new names?

```js
bnb.choice(
  hexColor1,
  hexColor2,
  rgbColor,
  rgbaColor
);
```

minamorl commented:
@wavebeem 
I corrected most of codes as you reviewed. But there's still a necessity of `bnb.eof`. If I run the codes without it, the example 2 and 3 won't works. Am I missing something?

BTW:
- I like the design of bnb. I confused the difference between `parser.next` and `parser.and` but it's okay once I get used to. The documentation was very helpful.
- I want to add a `parser.repeat(n: number)` function, which I implemented in a first commit.

wavebeem commented:
> But there's still a necessity of bnb.eof

I figured out the bug. hexColor2 should come before hexColor1.

```js
bnb.choice(hexColor1, hexColor2, rgbColor, rgbaColor)
```

hexColor1 and hexColor2 start out the same, so hexColor2 has to come first since it's longer.

> I like the design of bnb.
> The documentation was very helpful.

Yay! That's great :)

> I want to add a parser.repeat(n: number)

This could be a good new issue. Parsimmon has some functions like repeat that are helpful sometimes. I want to have discussions about new functions in GitHub issues so bnb doesn't get too big.

minamorl commented:
I've fixed that code a bit as you pointed out.

wavebeem commented:
@minamorl Does this new documentation help? https://wavebeem-bread-n-butter.netlify.app/api/#bnb.choice

minamorl commented:
It helps. Thanks. You can merge this at any time you like.
